File: src/saveUniverse.js
// Problem: https://codingcompetitions.withgoogle.com/codejam/round/00000000000000cb/0000000000007966
const readline = require("readline");
const rl = readline.createInterface(process.stdin, process.stdout);

// keeps track of the test case being solved
const caseTracker = {
  problems: [],
  results: [],
  numberOfLines: null,
  currentLine: null,
  nextLine() {
    this.currentLine = this.currentLine === null ? 0 : this.currentLine + 1;
  },
  setNumberOfLines(val) {
    this.numberOfLines = val;
  },
  addProblem(problem) {
    this.problems = [...this.problems, problem];
  },
  addResult(result) {
    this.results = [...this.results, result];
  }
};

// helper
function pipe(...fns) {
  return function folder(arg) {
    return fns.reduce((prev, fn) => fn(prev), arg);
  };
}

// functions to solve the problem
// applies damage to shield and returns new shield and current damage
function applyDamage([shield, laser]) {
  return [shield - laser, laser];
}
// amplifies damange and returns the shield and new damage
function amplifyDamage([shield, laser]) {
  return [shield, laser * 2];
}
// map string commands to functions
const instructionsMap = {
  C: amplifyDamage,
  S: applyDamage
};

const getFunctions = command => instructionsMap[command];

// swaps CS patters starting from the right
// we start from the right since those have higher power of two
function swap(commands) {
  const pivot = commands.lastIndexOf("CS");
  if (pivot === -1) {
    return commands;
  }
  return commands.slice(0, pivot).concat("SC", commands.slice(pivot + 2));
}

const initialDamage = 1;

// recusively hack the commands
// if swap returns the same and final shield is less than 0, we are DOOMED
// otherwise returns how many steps it took to hack the commands
function hackCommands(commands, shield, step = 0) {
  const robotFunctions = commands.split("").map(getFunctions);

  const [finalShield] = pipe(...robotFunctions)([shield, initialDamage]);

  if (finalShield < 0) {
    const newCommands = swap(commands);
    if (newCommands === commands) {
      return "IMPOSSIBLE";
    }
    return hackCommands(newCommands, shield, step + 1);
  }

  return step;
}

rl.on("line", function(line) {
  //code goes here
  caseTracker.nextLine();
  const lineNumber = caseTracker.currentLine;

  // for the first line, which specifies the number of cases
  if (!lineNumber) {
    return caseTracker.setNumberOfLines(parseInt(line));
  }

  caseTracker.addProblem(line);

  if (lineNumber === caseTracker.numberOfLines) {
    return rl.close();
  }
}).on("close", function() {
  caseTracker.problems.forEach((line, index) => {
    const [stringShield, commands] = line.split(" ");

    const shield = parseInt(stringShield);

    const result = hackCommands(commands, shield);

    caseTracker.addResult(`Case #${index + 1}: ${result}`);
  });

  caseTracker.results.forEach(result => console.log(result));

  process.exit(0);
});

File: src/saveUniverse.rs
// https://codingcompetitions.withgoogle.com/codejam/round/00000000000000cb/0000000000007966
use std::io;

type Res<T> = Result<T, Box<dyn std::error::Error>>;

fn nxt() -> String {
    let mut input = String::new();
    match io::stdin().read_line(&mut input) {
        Ok(_) => input,
        _ => panic!("Error reading line"),
    }
}

fn ptc<T: std::str::FromStr>() -> T {
    match nxt().trim().parse::<T>() {
        Ok(n) => n,
        _ => panic!("Error parsing"),
    }
}

fn apply_dmg(shield: i32, laser: i32) -> [i32; 2] {
    [shield - laser, laser]
}

fn amp_dmg(shield: i32, laser: i32) -> [i32; 2] {
    [shield, laser * 2]
}

fn swap(commands: &str) -> String {
    // any combination of CS is dangerous
    // it will double the damage and shoot
    // those CS at the end of thr string are the worst!
    // reverse the commands, find SC's on the reversed
    // replace them with CS and return the commands again
    commands
        .chars()
        .rev()
        .collect::<String>()
        .replacen("SC", "CS", 1)
        .chars()
        .rev()
        .collect::<String>()
}

fn hack(commands: &str, shield: i32, step: i32) -> i32 {
    // calculate final shield value by folding over the commands
    let final_shield: i32 = *commands
        .chars()
        .fold([shield, 1], |acc, val| match val {
            'C' => amp_dmg(acc[0], acc[1]),
            'S' => apply_dmg(acc[0], acc[1]),
            _ => [-1, -1],
        })
        .get(0)
        .expect("Problem with final shield");
    // if the final shield is less than 0
    if final_shield < 0 {
        // swap commands
        let swapped_commands: String = swap(&commands);
        // if the swap returns the same command string
        if &swapped_commands == commands {
            // return -1
            return -1;
        }
        // otherwise call hack again, with the new commands
        // the original shield
        // and the step increased by one
        return hack(&swapped_commands, shield, step + 1);
    }
    // if the final shield is more than 0, return the current step
    return step;
}

fn main() -> Res<()> {
    let n = ptc::<i32>();
    for case in 1..=n {
        let next = nxt();
        let mut it = next.split_whitespace();
        let shield = it.next().unwrap().parse::<i32>()?;
        let commands = it.into_iter().collect::<String>();

        match hack(&commands, shield, 0) {
            -1 => println!("Case #{}: IMPOSSIBLE", case),
            res => println!("Case #{}: {}", case, res),
        }
    }
    Ok(())
}
